### PR TITLE
Use the firewall role and the selinux role from the cockpit role

### PR DESCRIPTION
**Pull request by nhosoi**
- Introduce cockpit_manage_firewall to use the firewall role to manage the cockpit service. Default to false - means the firewall role is not used.

- Introduce cockpit_manage_selinux to use the selinux role to manage the ports in the cockpit service. Assign websm_port_t to the cockpit service ports. Default to false - means the selinux role is not used.

- Add the test check task tasks/check_firewall_selinux.yml for verify the ports status.

- Add meta/collection-requirements.yml.

### CODE REVIEW

**Comment by nhosoi** [comment]
[citest]

**Comment by nhosoi** [comment]
I forced to push a commit to update README.md after I set `[citest]`. Unfortunately, it prevents to display the CI test results in the CI test result section... But thanks to `check_jenkins.py` implemented by @richm . We could see all the tests successfully passed in the output.
```
TaskID   Started At          Role            PR  Platform               Duration Queue Time Status    
17297    2022-09-28T15:15:37 cockpit         76  RHEL-9.2/ansible-2.13  1713     14         PASSED    
17296    2022-09-28T15:15:37 cockpit         76  RHEL-9.2/ansible-2.9   1683     14         PASSED    
17295    2022-09-28T15:15:35 cockpit         76  RHEL-8.8/ansible-2.13  1779     13         PASSED    
17294    2022-09-28T15:15:35 cockpit         76  RHEL-8.8/ansible-2.9   1819     13         PASSED    
17293    2022-09-28T15:15:32 cockpit         76  RHEL-7.9/ansible-2.13  1817     10         PASSED    
17292    2022-09-28T15:15:32 cockpit         76  RHEL-7.9/ansible-2.9   1769     10         PASSED    
17291    2022-09-28T15:15:30 cockpit         76  CentOS-8/ansible-2.13  2714     7          PASSED    
17290    2022-09-28T15:15:30 cockpit         76  CentOS-8/ansible-2.9   2732     7          PASSED    
17289    2022-09-28T15:15:27 cockpit         76  CentOS-7/ansible-2.13  4243     5          PASSED    
17288    2022-09-28T15:15:27 cockpit         76  CentOS-7/ansible-2.9   4226     10         PASSED    
17287    2022-09-28T15:15:22 cockpit         76  Fedora-36/ansible-2.13 3120     5          PASSED    
17286    2022-09-28T15:15:22 cockpit         76  Fedora-36/ansible-2.9  3143     5          PASSED    
17285    2022-09-28T15:15:17 cockpit         76  Fedora-35/ansible-2.13 3128     0          PASSED    
17284    2022-09-28T15:15:17 cockpit         76  Fedora-35/ansible-2.9  3081     0          PASSED    
17283    2022-09-28T15:15:12 cockpit         76  RHEL-9.2/ansible-2.13  1713     19         PASSED    
17282    2022-09-28T15:15:12 cockpit         76  RHEL-9.2/ansible-2.9   1806     19         PASSED    
17281    2022-09-28T15:15:07 cockpit         76  RHEL-8.8/ansible-2.13  1845     15         PASSED    
17280    2022-09-28T15:15:07 cockpit         76  RHEL-8.8/ansible-2.9   1794     15         PASSED    
17279    2022-09-28T15:15:02 cockpit         76  RHEL-7.9/ansible-2.13  1809     10         PASSED    
17278    2022-09-28T15:15:02 cockpit         76  RHEL-7.9/ansible-2.9   1801     10         PASSED    
17277    2022-09-28T15:14:59 cockpit         76  CentOS-8/ansible-2.13  2738     7          PASSED    
17276    2022-09-28T15:14:59 cockpit         76  CentOS-8/ansible-2.9   2707     7          PASSED    
17275    2022-09-28T15:14:57 cockpit         76  CentOS-7/ansible-2.13  4292     10         PASSED    
17274    2022-09-28T15:14:57 cockpit         76  CentOS-7/ansible-2.9   4396     10         PASSED    
17273    2022-09-28T15:14:52 cockpit         76  Fedora-36/ansible-2.13 3191     5          PASSED    
17272    2022-09-28T15:14:52 cockpit         76  Fedora-36/ansible-2.9  3150     5          PASSED    
17271    2022-09-28T15:14:47 cockpit         76  Fedora-35/ansible-2.13 3179     0          PASSED    
17270    2022-09-28T15:14:47 cockpit         76  Fedora-35/ansible-2.9  3142     0          PASSED    
```

File: README.md [hunk 1]
```diff
@@ -81,17 +96,35 @@ Configure settings in the /etc/cockpit/cockpit.conf file.  See [`man cockpit.con
...
+Default to `false`.
+
+NOTE: `cockpit_manage_firewall` is limited to *adding* ports.
+It cannot be used for *removing* ports.
+If you want to remove ports, you will need to use the firewall system
+role directly.
 
-    semanage port -m -t websm_port_t -p tcp 443
+NOTE: This functionality is supported only when the managed host's `os_family`
+is `RedHat`.
```
Comment (martinpitt) [inline README.md:111]: Why wouldn't that work on Fedora? That even gets mentioned explicitly a few lines up.

File: README.md [hunk 2]
```diff
@@ -81,17 +96,35 @@ Configure settings in the /etc/cockpit/cockpit.conf file.  See [`man cockpit.con
...
 
-for ports that are already defined in the SELinux policy, such as 443, or
+    cockpit_manage_selinux: false
+Boolean flag allowing to configure selinux using the selinux role.
+Assign `websm_port_t` to the port.
```
Comment (martinpitt) [inline README.md:115]: This is too generic IMHO. Proposal:

> The default SELinux policy does not allow Cockpit to listen to anything else than port 9090. If you change the port, enable this to use the `selinux` role to set the correct port permissions (`websm_port_t`).

File: README.md [hunk 3]
```diff
@@ -81,17 +96,35 @@ Configure settings in the /etc/cockpit/cockpit.conf file.  See [`man cockpit.con
...
+If the variable is set to false, the `cockpit role` does not manage the
+selinux
```
Comment (martinpitt) [inline README.md:117]: "the selinux" is ungrammatical. "SELinux" or "the `selinux` role", or perhaps "SELinux permissions of the cockpit port".

File: tasks/selinux.yml [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+- name: Ensure the service and the ports status with the selinux role
+  include_role:
+    name: fedora.linux_system_roles.selinux
+  vars:
+    _cockpit_port: "{{ cockpit_port if cockpit_port is defined else 9090 }}"
+    selinux_ports: "{{ [{'ports': _cockpit_port, 'proto': 'tcp',
+                         'setype': 'websm_port_t',
+                         'state': 'present', 'local': 'true'}] }}"
+  when:
+    - cockpit_manage_selinux | bool
```
Comment (martinpitt) [inline tasks/selinux.yml:12]: I'd add an additional condition that `cockpit_port is defined` (and remove that from line 7), so that this isn't even invoked when the port doesn't get customized (the common case). Avoids traps and is faster.

File: tasks/firewall.yml [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+    name: fedora.linux_system_roles.firewall
+  vars:
+    _cockpit_port: "{{ cockpit_port if cockpit_port is defined else 9090 }}"
+    _cockpit_port_proto: "{{ _cockpit_port }}/tcp"
+    firewall: "{{ [{'service': 'cockpit', 'state': 'enabled'}]
+                  if (_cockpit_port | int) == 9090 else
+                  [{'port': _cockpit_port_proto, 'state': 'enabled'}] }}"
+  when:
+    - cockpit_manage_firewall | bool
+    - ansible_facts['os_family'] == 'RedHat'
```
Comment (martinpitt) [inline tasks/firewall.yml:14]: See above -- this ought to work perfectly fine on Fedora as well?

File: tests/tests_certificate.yml [hunk 6]
```diff
@@ -23,7 +23,12 @@
   vars:
     cockpit_packages: minimal
   roles:
-    - linux-system-roles.cockpit
+    - role: linux-system-roles.cockpit
+      public: true
```
Comment (martinpitt) [inline tests/tests_certificate.yml:27]: This `public` is an unrelated change, and should be a separate commit. Same for the other instances below.

File: tests/tests_certificate_runafter.yml [hunk 7]
```diff
@@ -84,6 +85,9 @@
           assert:
             that: "'status: MONITORING' in result.stdout"
 
+        - name: test - ensure cockpit_port is configured for firewall and selinux
+          include_tasks: tasks/check_firewall_selinux.yml
```
Comment (martinpitt) [inline tests/tests_certificate_runafter.yml:89]: Please don't add this to *all* tests -- separation of concerns, this is unnecessary redundancy. The one place where it really matters is `tests_port.yml`, and perhaps `tests_default.yml` in addition, but none of the others.

File: tests/tests_default.yml [hunk 8]
```diff
@@ -42,5 +43,8 @@
           register: result
           failed_when: result.stat.exists
 
+        - name: test - ensure cockpit_port is configured for firewall and selinux
+          include_tasks: tasks/check_firewall_selinux.yml
```
Comment (martinpitt) [inline tests/tests_default.yml:47]: Note that the default Fedora and RHEL configuration is fine by itself -- this should *not* mangle anything in the default case. The tasks file does not do anything harmful, so it's fine to call it here; I just wanted to point it out.

**Review by martinpitt — CHANGES_REQUESTED** [review]
Thanks @nhosi! Test-wise you got a lot further than me, I'm happy for you to take over, and I'll close my PR #73. I have several small comments, and I'd really appreciate if you could split this up in to separate commits (one for `public`, one for firewalld, one for selinux). Cheers!

File: README.md [hunk 9]
```diff
@@ -8,6 +8,21 @@ Installs and configures the Cockpit Web Console for distributions that support i
...
+    (Please see also `cockpit_manage_firewall` and `cockpit_manage_selinux`
+     in [`Role Variables`](#role-variables).
+
+    If the `cockpit` is a role from the `fedora.linux_system_roles`
+    collection or from the Fedora RPM package, the requirement is already
+    satisfied.
+
+    Otherwise, please run the following command line to install the collection.
+    ```
+    ansible-galaxy collection install -r meta/collection-requirements.yml
```
[the remark this answers is not in the document]
Comment (richm) [reply README.md:23]: > This would only work if you are in a git checkout of this repo, right?

It is also present in the meta directory of the role if you install from Galaxy.

> Perhaps this can be spelled out instead of the "Otherwise"?

I'm assuming it is implied from the context, but if you can suggest something that isn't too verbose, we can change this and all of the other roles where we already have this exact same verbiage.

File: README.md [hunk 10]
```diff
@@ -81,17 +96,35 @@ Configure settings in the /etc/cockpit/cockpit.conf file.  See [`man cockpit.con
     cockpit_port: 9090
 Cockpit runs on port 9090 by default. You can change the port with this option.
 
-Note that the default SELinux policy does not allow Cockpit to listen to anything else than port 9090, so you need to allow that first, with e.g.
+    cockpit_manage_firewall: false
```
[the remark this answers is not in the document]
Comment (richm) [reply README.md:99]: We really should be using `true` and `false` everywhere - https://github.com/redhat-cop/automation-good-practices/blob/main/coding_style/README.adoc#yaml-and-jinja2-syntax "Use true and false for boolean values in playbooks."

but yes, it will look strange here if we are not consistent, but we may want to convert everything in the future - note that some of the other roles have already done this - https://github.com/linux-system-roles/ha_cluster/pull/60

File: tasks/selinux.yml (hunk 4 above)
Comment (richm) [reply tasks/selinux.yml:12]: hmm - since `cockpit_port` is part of the public api of the role, it should be defined in `defaults/main.yml` as `null` - so check for `cockpit_port is not none` (`null` in YAML == `none` in Jinja)

File: tasks/firewall.yml (hunk 5 above)
Comment (richm) [reply tasks/firewall.yml:14]: In ansible, `ansible_facts['os_family']` `RedHat` includes `Fedora`, `CentOS`, and `RedHat` (and maybe in the future things like `AlmaLinux`, `RockyLinux`, et. al.)

File: README.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (martinpitt) [reply README.md:23]: If it is implied for users who are more familiar with this, then fine of course. I was just bringing it up as it sounds confusing to me -- this relative file path *definitively* makes some assumption of what directory you are in. That is fine for a git clone, but this README also ends up in the official documentation, I suppose.

File: tasks/firewall.yml (hunk 5 above)
Comment (martinpitt) [reply tasks/firewall.yml:14]: I see, thanks!

File: tests/tests_certificate.yml (hunk 6 above)
Comment (richm) [reply tests/tests_certificate.yml:27]: I haven't checked, but it looks like check_firewall_selinux.yml requires role private variables in order to check the values? If so, then I think `public: true` is justified

File: README.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (richm) [reply README.md:23]: > If it is implied for users who are more familiar with this, then fine of course. I was just bringing it up as it sounds confusing to me -- this relative file path _definitively_ makes some assumption of what directory you are in.

It might also imply "the `meta` directory of the role source from wherever you are using the role".  Again, if you can suggest something more appropriate, we will consider it.

> That is fine for a git clone, but this README also ends up in the official documentation, I suppose.

Depends on what you mean by "official documentation".  The collection install is *only* needed if you are using the role from Galaxy or github.  If you are using the role from any other context - from the collection as fedora.linux_system_roles.cockpit - from any of the RPM distributions - from Automation Hub - you *do not* need to do this step.

[the remark this answers is not in the document]
Comment (martinpitt) [reply README.md:23]: Ack, thanks @richm for the clarification! Makes sense then.

**Comment by nhosoi** [comment]
> preciate if you could split this up in to separate commits (one for `public`, one for firewalld, one for selinux). Cheers!

How about the test script, `tests/tasks/check_firewall_selinux.yml`? Do you want to make it into two files; one for firewall and another for selinux?

File: tests/tests_certificate.yml (hunk 6 above)
Comment (nhosoi) [reply tests/tests_certificate.yml:27]: Right. `check_firewall_selinux.yml` needs it. And I'm going to drop the unnecessary `public: true` other than `tests_config.yml`, `tests_packages_full.yml`, and `tests_port.yml` along with `check_firewall_selinux.yml`.

File: tasks/selinux.yml (hunk 4 above)
Comment (nhosoi) [reply tasks/selinux.yml:12]: > hmm - since `cockpit_port` is part of the public api of the role, it should be defined in `defaults/main.yml` as `null` - so check for `cockpit_port is not none` (`null` in YAML == `none` in Jinja)

Is it ok for me to make this change in this pr?

File: README.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (nhosoi) [reply README.md:99]: Ok. `yes/no` for cockpit for now. Let me update README and tests.
I assume the change is not needed in the files in tasks and defaults... Right?

File: tests/tests_certificate.yml (hunk 6 above)
Comment (martinpitt) [reply tests/tests_certificate.yml:27]: @nhosoi : I see -- I didn't realize that it's really the new tasks/* files which require `public:`. That's fine then.

**Comment by martinpitt** [comment]
@nhosoi : a new feature and a test should always be in the same commit, as they belong together. Same for the documentation. The point is not to make commits as small as possible, but only do one thing/feature/bug fix/change per commit (*including* the corresponding docs, tests, etc.). That makes them easier to understand/review, possible to bisect, and backport.

I leave it up to you whether you want to split tests/tasks/check_firewall_selinux.yml; IMHO the functionality belongs together, and I'd even rename it to `check_port.yml`. Then the "firewall" commit would do the firewalld check there, and the "selinux" commit would add the web_sm_t check to the same file.

File: tasks/selinux.yml (hunk 4 above)
Comment (richm) [reply tasks/selinux.yml:12]: > > hmm - since `cockpit_port` is part of the public api of the role, it should be defined in `defaults/main.yml` as `null` - so check for `cockpit_port is not none` (`null` in YAML == `none` in Jinja)
> 
> Is it ok for me to make this change in this pr?

Yes, please.

File: README.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (richm) [reply README.md:99]: > Ok. `yes/no` for cockpit for now. Let me update README and tests. I assume the change is not needed in the files in tasks and defaults... Right?

Please use `yes` and `no` in all of the code you are adding.

**Comment by nhosoi** [comment]
[citest]

**Comment by nhosi** [comment]
> Thanks @nhosi! Test-wise you got a lot further than me, I'm happy for you to take over, and I'll close my PR #73. I have several small comments, and I'd really appreciate if you could split this up in to separate commits (one for `public`, one for firewalld, one for selinux). Cheers!

Hi! You tagged a wrong GitHub account. I will unsubscribe this chat. Br, @nhosi

**Comment by nhosoi** [comment]
[citest]

**Comment by nhosoi** [comment]
[citest]

**Comment by nhosoi** [comment]
[citest]

File: tests/tasks/check_port.yml [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+              - policycoreutils-python
+            state: present
+          when: ( ansible_python_version is version('3', '<') and
+                  ansible_distribution in ["Fedora", "CentOS", "RedHat", "Rocky"] )
+
+        - name: Install SELinux python3 tools
+          package:
+            name:
+              - libselinux-python3
+              - policycoreutils-python3
```
Comment (richm) [inline tests/tasks/check_port.yml:37]: looks like `semanage` on Fedora is provided by `policycoreutils-python-utils`

**Comment by richm** [comment]
ok - looks good - please rebase on top of the latest master branch and we can merge

**Comment by nhosoi** [comment]
> ok - looks good - please rebase on top of the latest master branch and we can merge

Thank you, @richm. Done!

**Comment by richm** [comment]
spoke too soon - looks like merge is blocked - @martinpitt needs to review, and clear the requested changes flag.

**Review by martinpitt — APPROVED** [review]
Very cool, thanks @nhosoi !
